### DISCUSSION TITLE
싱글톤 패턴에 대해서 설명해주세요

### DISCUSSION THREAD

**Discussion by alstn113** [2024-woowacourse-study/level-interview#79] [OOP]
.

**Comment by jimi567** [comment]
싱글톤 패턴은 어떤 클래스가 최대 하나의 인스턴스만을 갖도록 보장하는 패턴이다. 싱글톤 패턴을 사용하게 되면, 반복적으로 사용될 수 있는 클래스 객체에 대해 같은 인스턴스를 사용하도록 하기 때문에 자원을 효율적으로 사용할 수 있다.
대게, 싱글톤 객체를 얻어오는 메서드명은 getInstance를 사용하는 것이 컨벤션이다.

(사용 경험은 저의 경험이니깐 구어체로 작성할게요~ 면접 연습 할 때는 자기 사용 경험 얘기해주세요!)

저는 싱글톤 패턴과 상태 패턴을 혼합하여 사용한 경험이 있습니다.
상태 패턴은 도메인의 상태를 객체화 하여, 확장성 및 유지보수 측면에서 큰 이점을 주는 디자인 패턴이지만, 객체 생성이 빈도가 높아서, GC를 자주 유발 할 수 있다는 점 등의 명확한 단점이 있습니다. 이때, 제가 상태 패턴으로 사용한 객체의 경우 가변적인 상태(멤버 변수)가 없었기 때문에, 싱글톤 패턴을 활용하면, 상태패턴의 단점을 해소할 수 있다 생각하여, 사용한 경험이 있습니다.

**Comment by nayonsoso** [comment]
싱글톤 패턴이란, 어플리케이션 전체에서 어떤 클래스의 인스턴스를 하나만 만들도록 하는 패턴을 말합니다. 

싱글톤 패턴은 일관성과 메모리 절약이라는 장점을 갖습니다.
싱글톤 패턴을 사용하면, 생성자를 여러번 호출해도 같은 인스턴스에 접근하게 됩니다. 
따라서 일관된 상태의 객체를 접근할 수 있습니다.
또한 인스턴스를 하나만 만들기 때문에 메모리를 절약할 수 있습니다.

싱글톤 패턴을 사용할 때는 객체가 상태를 갖지 않게 하거나 불변의 상태를 갖게 해야 합니다.
만약 가변의 상태를 갖는다면, 여러개의 쓰레드가 값을 수정할 때 동시성 문제가 발생할 수 있습니다.

**Comment by le2sky** [comment]
체스 미션에서 널 객체인 EmptyPiece에 대한 싱글톤 객체를 만들었습니당

```java
class EmptyPiece extends AbstractPiece {
    ...
    private static final Piece INSTANCE = new EmptyPiece();
    ..
```

**Comment by GIVEN53** [comment]
하나의 인스턴스
일관성, 메모리 절약
가변 상태 x -> 동시성 문제
